### ISSUE TITLE
Permissions for changing and deleting pages

### ISSUE THREAD

**Issue by fvsch** [getgrav/grav-plugin-admin#512]
It would be excellent if we could lock down some pages so that users can only edit specific pages, not the whole site. Maybe something on top of user groups?

For instance there could be default permissions, perhaps in `user/config/admin/admin.yaml` (or maybe it would make more sense in blueprints, which allows setting this config depending on template):

``` yaml
# Only admins can delete pages
permissions:
  pages:
    write: [admin, editor]
    delete: [admin]
```

And YAML config in each page could overwrite these default settings:

``` yaml

---
title: Awesome page
permissions:
  write: [admin] # Limit changes to the admin group in the admin plugin
  delete: [] # Protect this page from deletion from the admin plugin

---
```
## Use cases for locking down a page
- If your site structure rests on a few "container" pages (blog, section1, section2…) and you want to avoid careless deletions of _all the content_ in a section by locking the level 1 pages.
- If you want to protect the error page or other "technical" pages.
- A business communication website where one editor or group of editors should be able to create new blog posts, perhaps modify the root "blog" page but NOT delete it, and should not be able to modify (let alone delete) other pages in other sections (or the home page).
## Permission types

I think a simple implementation could be limited to 2 permissions:
- write
- delete

Excluding these other common permission types (less useful or harder to manage):
- read (the need to hide content to admin users seems low)
- page ownership (write/delete your "own" pages) -> adds a big level of complexity
- addchild (controlling whether a user group can add a child page inside a page) -> adds some complexity too

**Comment by rhukster** [comment]
Some good ideas here.  I actually already had a note with some ideas on how to improve permissions. I'll make sure this ticket is taken into account also.  Thanks!

**Comment by boettges** [comment]
@rhukster Hey, just wanted to know if there is any progress on that topic that I missed? 

For a specific project I need group-based edit-permissions for certain pages or pages of a folder. It's the only requirement that could potentially keep us from using Grav 😞

It would be really helpful, if you could point in the right direction, researching this.
Thanks a lot in advance for your help.

**Comment by jimblue** [comment]
Most wanted feature!

**Comment by Aisbergg** [comment]
bump

**Comment by brianjschott** [comment]
I have students who want to work on our school's website. Right now they are just passing me the information to post on the blog; it would be helpful to be able to give them access to specific pages so I don't have to worry about a student accidentally or purposefully editing the frontpage or something else important.

Thanks - hope that progress is being made on this feature!!

**Comment by ocean123** [comment]
When can we aspect this feature. I hope so that it will be in the next version. :-))

**Comment by bleutzinn** [comment]
@brianjschott You could checkout the [Editable with SimpleMDE Plugin](https://github.com/bleutzinn/grav-plugin-editable-simplemde#editable-with-simplemde-plugin) and test whether or not it suits your purpose. The latest version supports page based permissions using the Grav user and roles system. So you can create groups as well. Editing is limited to pages and in the frontend only.

**Comment by robinmoisson** [comment]
I have to give editorial access to non-technical users, this would definitely help me sleep at night ! :+1:

**Comment by Sogl** [comment]
Any progress on this? Most wanted feature.

**Comment by mahagr** [comment]
I'm currently working on Flex Pages support and I already have permissions for deleting pages (though right now without owner support).

**Comment by Sogl** [comment]
> Some good ideas here. I actually already had a note with some ideas on how to improve permissions. I'll make sure this ticket is taken into account also. Thanks!

@rhukster over 3 years have passed from your comment. Some of my clients really want this feature, but we as Grav loving developers still no info about it.

@mahagr What's the Flex Pages? It's about css flexbox or something different? Please give more info 😉

**Comment by mahagr** [comment]
Flex Pages are based on Grav Flex Objects, which is just a cool name for the classes (you can find them in `Grav\Framework\Flex`. :)

In short, Flex Objects will be replacing Flex Directory plugin; the core classes are in Grav itself, but for CRUD you will need the Flex Objects plugin, which will provide listing and editing views which will work in both admin and in the frontend Though that said, I am not sure if it will ship with working frontend editing or if you need to do some custom classes/routing to get it to work. Anyways, this plugin is only needed for admin tasks, Flex itself runs without it.

Flex Pages in implementation of the current Grav Pages, but based on these new Flex classes. Using Flex has a few benefits over the old solution:

* it keeps an index from all the objects and knows which objects have been updated and when
* because of the index, objects are only being loaded on demand
* it has built-in caching for collection searches, method calls and even rendering
* collections are far more powerful than with old pages; you can search almost anything
* `FlexPage` is self-contained and much easier to use than `Page`
* Flex allows you to easily customize pages for your needs
* It allows you to create own admin section for example for blog posts -- separated from the other pages
* ...

There are even more benefits, but we have been using these already in some of our projects and everyone is so excited about this. We get easily 10 times better performance on 4000-page site. We can create custom types, even pages that are stored separately from the regular pages and are in a separate section inside admin. Users can create their own pages by filling simple forms etc...

**Comment by paulhibbitts** [comment]
Thanks so much for sharing this info @mahagr 👍 Of everything you mention, more advanced page collections, page customization and separate Admin section for different page types is most interesting to me.

**Comment by localnetwork** [comment]
Following this post. Badly need this right now.

![60153027-8420fb80-9815-11e9-831b-7a32f4f48f7b](https://user-images.githubusercontent.com/15809239/63492737-a3be6280-c4ec-11e9-8173-596535b9d045.jpg)

**Comment by paulrudy** [comment]
I would really appreciate this function as well. Related to this [open enhancement issue](https://github.com/getgrav/grav-plugin-admin/issues/777) as well

**Comment by mahagr** [comment]
This has already been implemented in Grav 1.7 / Admin 1.10
